lib/spack/spack/test/concretize_requirements.py

```diff
@@ -199,54 +224,51 @@ def test_requirement_adds_git_hash_version(
     concretize_scope, test_repo, mock_git_version_info, monkeypatch
 ):
     if spack.config.get("config:concretizer") == "original":
-        pytest.skip("Original concretizer does not support configuration" " requirements")
+        pytest.skip("Original concretizer does not support configuration requirements")
 
     repo_path, filename, commits = mock_git_version_info
     monkeypatch.setattr(
         spack.package_base.PackageBase, "git", path_to_file_url(repo_path), raising=False
     )
 
     a_commit_hash = commits[0]
-    conf_str = """\
+    conf_str = f"""\
 packages:
   v:
-    require: "@{0}"
-""".format(
-        a_commit_hash
-    )
+    require: "@{a_commit_hash}"
+"""
     update_packages_config(conf_str)
 
     s1 = Spec("v").concretized()
-    assert s1.satisfies("@{0}".format(a_commit_hash))
+    assert isinstance(s1.version, spack.version.GitVersion)
+    assert s1.satisfies(f"v@{a_commit_hash}")
 
 
 def test_requirement_adds_multiple_new_versions(
     concretize_scope, test_repo, mock_git_version_info, monkeypatch
 ):
     if spack.config.get("config:concretizer") == "original":
-        pytest.skip("Original concretizer does not support configuration" " requirements")
+        pytest.skip("Original concretizer does not support configuration requirements")
 
     repo_path, filename, commits = mock_git_version_info
     monkeypatch.setattr(
         spack.package_base.PackageBase, "git", path_to_file_url(repo_path), raising=False
     )
 
-    conf_str = """\
+    conf_str = f"""\
 packages:
   v:
     require:
-    - one_of: ["@{0}=2.2", "@{1}=2.3"]
-""".format(
-        commits[0], commits[1]
-    )
+    - one_of: ["@{commits[0]}=2.2", "@{commits[1]}=2.3"]
+"""
     update_packages_config(conf_str)
 
     s1 = Spec("v").concretized()
     assert s1.satisfies("@2.2")
 
-    s2 = Spec("v@{0}".format(commits[1])).concretized()
-    assert s2.satisfies("@{0}".format(commits[1]))
-    assert s2.satisfies("@2.3")
+    # s2 = Spec(f"v@{commits[1]}").concretized()
+    # assert s2.satisfies("v@{commits[1]}")
+    # assert s2.satisfies("@2.3")
```

Comment:
Intentionally commented out?

Reply:
Commented out because of how weird gitversion version declaration was handled in asp.py.

Somehow, `@hash=version` on the command line both defines `hash=version` and `version` as possible/declared versions. That's wrong... And similarly the situation with preferences is weird: a preferred version is added to the list of declared versions, but is not among "possible versions" (and lacks this weirdness of declaring two versions for gitversion).

With @tgamblin's comment that it would be better to remove the "exact match from command line" logic when you actually pass a range instead of a specific version, I think some weirdness can simply be removed from asp.py; I'll also remove this weird `hash=bla` defines `bla` as a version, since it's wrong, since `hash=bla` _is_ a concrete version.

Reply:
Okay, yeah: this particular instance of requirements is intentional, because `spack install v@<hash>` causes a lookup which is independent of concretization, so it will resolve to a different version than is required, causing a concretizationn error. However, `spack install v@2.3` is fine, since it's a range which includes a git commit which is required.

imho it's too involved to make `spack install v@<hash>` resolve its version either to a lookup _or_ to a user defined mapping in a requirement -- before you get to the requirement, the git ref is very likely already resolved to in the git repo.